### ISSUE TITLE
Dockerize indicator runners

### ISSUE THREAD

**Issue by melange396** [cmu-delphi/covidcast-indicators#1968]
Following up to #1967, for consistency and reproducibility, we should "Dockerize" (or equivalent/similar) our indicator runtime environments.  Though it will not be perfect, it will help us run the same code on different machines without having to worry about subtle differences in configurations or versioning of dependencies.

Our current installations essentially run on "bare hardware" (not even inside `venv`s, AFAICT) where different jobs may expect particular setups but instead actually be limited by each other's constraints.  This will be a kind of paradigm shift in that our deployment processes and job scheduling/triggering will have to change.

Somewhat related to https://github.com/cmu-delphi/delphi-epidata/issues/1389 .

**Comment by dshemetov** [comment]
Good idea! Started thinking a little about this (using some tips from [this blog](https://pythonspeed.com/articles/base-image-python-docker-images/)), here's something for the hhs indicator to get started (chose it because I don't think it requires extra credentials). The lint command worked, haven't tried running the indicator yet:

```sh
# == covidcast-indicators/hhs_hosp/Dockerfile ==
FROM python:3.8.19-slim-bookworm

RUN mkdir /usr/src/app
WORKDIR /usr/src/app

RUN apt-get update
RUN apt-get install -y make

COPY . /usr/src/app

RUN make install
# == covidcast-indicators/hhs_hosp/Dockerfile ==

# == covidcast-indicators/hhs_hosp/.dockerignore ==
# Ignore bulky directories we bind-mount
cache
receiving
# Ignore local virtual environment
env
# == covidcast-indicators/hhs_hosp/.dockerignore ==

# Commands to be run in the covidcast-indicators/hhs_hosp directory
docker build -f Dockerfile . -t delphi_hhs
docker run -d \
  -it \
  --mount type=bind,source="$(pwd)"/cache,target=/usr/src/app/cache \
  --mount type=bind,source="$(pwd)"/receiving,target=/usr/src/app/receiving \
  -e PARAM=value \
  delphi_hhs make lint
docker run -d \
  -it \
  --mount type=bind,source="$(pwd)"/cache,target=/usr/src/app/cache \
  --mount type=bind,source="$(pwd)"/receiving,target=/usr/src/app/receiving \
  -e PARAM=value \
  delphi_hhs python -m delphi_hhs
```

Next steps might be something like:

* make an analogue for an indicator that's pulling current data (hhs is not, for now)
* get it going on staging and compare outputs

**Comment by dsweber2** [comment]
So, I have maybe a dumb patch of an idea that can temporarily make sure that indicators have up to date environments:

A chronicle job that backs up the `venv` folder, does `make clean; make install` on staging for each indicator, makes sure that works, and then after 1-3 days does the same on Prod (enough time to cancel if it broke on staging). Run this like once a month or something.

**Comment by melange396** [comment]
I was mistaken, we do actually make use of `venv`s for the indicators...  I thought it was necessary to execute the `activate` script to properly set up the environment, which we do not do in our scheduled job run ; in fact [it is not required](https://docs.python.org/3/library/venv.html#:~:text=You%20don%E2%80%99t%20specifically%20need%20to%20activate%20a%20virtual%20environment), and the way we invoke indicator jobs should take advantage of their respective virtual environments.

Using Jenkins (on a separate machine), we ["build" environments and tar them up](https://github.com/cmu-delphi/covidcast-indicators/blob/84d059751b646c0075f1a384741f2c1d80981269/jenkins/build-and-package.sh#L19-L34) and then [unzip those directory trees](https://github.com/cmu-delphi/covidcast-indicators/blob/84d059751b646c0075f1a384741f2c1d80981269/ansible/ansible-deploy.yaml#L7-L12) in the prod and staging machines.  However, such environments are [not intended to be moved](https://docs.python.org/3/library/venv.html#:~:text=Not%20considered%20as%20movable%20or%20copyable%20%E2%80%93%20you%20just%20recreate%20the%20same%20environment%20in%20the%20target%20location), [even to a different directory on the same machine](https://docs.python.org/3/library/venv.html#:~:text=If%20for%20any%20reason%20you%20need%20to%20move%20the%20environment%20to%20a%20new%20location%2C%20you%20should%20recreate%20it%20at%20the%20desired%20location%20and%20delete%20the%20one%20at%20the%20old%20location.).  Perhaps it is good that we do not "`activate`" the environments because there is path information from the build machine that is included in the script:
```sh
$ less ~indicators/runtime/nchs_mortality/env/bin/activate | grep nchs_mortality
VIRTUAL_ENV="/mnt/data/jenkins/workspace/covidcast-indicators_prod/nchs_mortality/env"
```
This approach has the "build once and then distribute" paradigm similar to Docker, but it unfortunately has these problems (and i am surprised we havent been bitten by them (yet?)).

After consulting w/ @korlaxxalrok , he thinks that (but dont quote me on this!) Jenkins can be made to build virtual environments on the prod/staging servers or Jenkins could build Docker images in a similar way instead.  He also suggested that we could get GH actions to do it, but voiced concerns about secrets being leaked from there (unless we are careful to use [methods to mask variables in the logs](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#masking-a-value-in-a-log)).
